datacube/api/grid_workflow.py

```diff
@@ -0,0 +1,459 @@
+        """
+        :rtype: datacube.model.DatasetType
+        """
+        return self.sources.values[0][0].product
+
+    def __getitem__(self, chunk):
+        sources = _fast_slice(self.sources, chunk[: len(self.sources.shape)])
+        geobox = self.geobox[chunk[len(self.sources.shape):]]
+        return Tile(sources, geobox)
+
+    # TODO(csiro) Split on time range
+    def split(self, dim: str, step: int = 1):
+        """
+        Splits along a non-spatial dimension into Tile objects with a length of 1 or more in the `dim` dimension.
+
+        :param dim: Name of the non-spatial dimension to split
+        :param step: step size to split
+        :return: tuple(key, Tile)
+        """
+        axis = self.dims.index(dim)
+        indexer = [slice(None)] * len(self.dims)
+        size = self.sources[dim].size
+        for i in range(0, size, step):
+            indexer[axis] = slice(i, min(size, i + step))
+            yield self.sources[dim].values[i], self[tuple(indexer)]
+
+    def split_by_time(self, freq: str = "A", time_dim: str = "time", **kwargs):
+        """
+        Splits along the `time` dimension, into periods, using pandas offsets, such as:
+        :
+            'A': Annual
+            'Q': Quarter
+            'M': Month
+        See: http://pandas.pydata.org/pandas-docs/stable/timeseries.html?highlight=rollback#timeseries-offset-aliases
+
+        :param freq: time series frequency
+        :param time_dim: name of the time dimension
+        :param kwargs: other keyword arguments passed to ``pandas.period_range``
+        :return: Generator[tuple(str, Tile)] generator of the key string (eg '1994') and the slice of Tile
+        """
+        # extract first and last timestamps from the time axis, note this will
+        # work with 1 element arrays as well
+        start_range, end_range = self.sources[time_dim].data[[0, -1]]
+
+        for p in pd.period_range(start=start_range, end=end_range, freq=freq, **kwargs):
+            sources_slice = self.sources.loc[
+                {time_dim: slice(p.start_time, p.end_time)}
+            ]
+            yield str(p), Tile(sources=sources_slice, geobox=self.geobox)
+
+    def __str__(self):
+        return "Tile<sources={!r},\n\tgeobox={!r}>".format(self.sources, self.geobox)
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class GridWorkflow:
+    """
+    GridWorkflow deals with cell- and tile-based processing using a grid defining a projection and resolution.
+
+    Use GridWorkflow to specify your desired output grid.  The methods :meth:`list_cells` and :meth:`list_tiles`
+    query the index and return a dictionary of cell or tile keys, each mapping to a :class:`Tile` object.
+
+    The :class:`.Tile` object can then be used to load the data without needing the index,
+    and can be serialized for use with the `distributed` package.
+    """
+
+    def __init__(
+        self,
+        index: Index,
+        grid_spec: GridWorkflow | None = None,
```

Comment:
Is this a copy constructor, or should `grid_spec` be called `grid_workflow`?

Reply:
Updated